File: src/main/java/com/rednorte/sigle/citas_service/service/CitaService.java
package com.rednorte.sigle.citas_service.service;

import com.rednorte.sigle.citas_service.repository.CancelacionRepository;
import com.rednorte.sigle.citas_service.repository.CitaRepository;
import com.rednorte.sigle.citas_service.config.RabbitMQConfig;
import com.rednorte.sigle.citas_service.model.Cancelacion;
import com.rednorte.sigle.citas_service.model.CanceladoPor;
import com.rednorte.sigle.citas_service.model.Cita;
import com.rednorte.sigle.citas_service.model.EstadoCita;
import com.rednorte.sigle.citas_service.model.Medico;

import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CitaService {

    private final CitaRepository citaRepository;
    private final CancelacionRepository cancelacionRepository;
    private final RabbitTemplate rabbitTemplate;
    private final MedicoService medicoService;

    @CircuitBreaker(name = "citasService", fallbackMethod = "fallbackGetAll")
    public List<Cita> getAll() {
        return citaRepository.findAll();
    }

    public List<Cita> fallbackGetAll(Exception e) {
        return List.of();
    }

    @CircuitBreaker(name = "citasService", fallbackMethod = "fallbackGetById")
    public Cita getById(Long id) {
        return citaRepository.findById(id).orElseThrow(() -> new RuntimeException("Cita no encontrada"));
    }

    public Cita fallbackGetById(Long id, Exception e) {
        throw new RuntimeException("Servicio de citas no disponible temporalmente");
    }

    @CircuitBreaker(name = "citasService", fallbackMethod = "fallbackGetByPacienteId")
    public List<Cita> getByPacienteId(Long pacienteId) {
        return citaRepository.findByPacienteId(pacienteId);
    }

    public List<Cita> fallbackGetByPacienteId(Long pacienteId, Exception e) {
        return List.of();
    }

    @CircuitBreaker(name = "citasService", fallbackMethod = "fallbackGetByPacienteIdPaginado")
    public Page<Cita> getByPacienteIdPaginado(Long pacienteId, Pageable pageable) {
        return citaRepository.findByPacienteId(pacienteId, pageable);
    }

    public Page<Cita> fallbackGetByPacienteIdPaginado(Long pacienteId, Pageable pageable, Exception e) {
        return Page.empty();
    }

    @CircuitBreaker(name = "citasService", fallbackMethod = "fallbackGetByMedicoId")
    public List<Cita> getByMedicoId(Long medicoId) {
        return citaRepository.findByMedicoId(medicoId);
    }

    public List<Cita> fallbackGetByMedicoId(Long medicoId, Exception e) {
        return List.of();
    }

    @CircuitBreaker(name = "citasService", fallbackMethod = "fallbackGetHorasOcupadas")
    public List<String> getHorasOcupadas(Long medicoId, LocalDate fecha) {
        LocalDateTime inicio = fecha.atStartOfDay();
        LocalDateTime fin = fecha.atTime(LocalTime.MAX);
        List<Cita> citas = citaRepository.findByMedicoIdAndFechaHoraBetweenAndEstadoNot(
            medicoId, inicio, fin, EstadoCita.CANCELADA);
        return citas.stream()
            .map(c -> c.getFechaHora().toLocalTime().toString())
            .collect(Collectors.toList());
    }

    public List<String> fallbackGetHorasOcupadas(Long medicoId, LocalDate fecha, Exception e) {
        return List.of();
    }

    @Transactional
    @CircuitBreaker(name = "citasService", fallbackMethod = "fallbackAgendarCita")
    public Cita agendarCita(Cita nuevaCita, Long medicoId) {
        Medico medico = medicoService.getById(medicoId);
        nuevaCita.setMedico(medico);
        nuevaCita.setEstado(EstadoCita.PROGRAMADA);
        return citaRepository.save(nuevaCita);
    }

    public Cita fallbackAgendarCita(Cita nuevaCita, Long medicoId, Exception e) {
        throw new RuntimeException("No se puede agendar la cita. Servicio no disponible temporalmente.");
    }

    @Transactional
    @CircuitBreaker(name = "citasService", fallbackMethod = "fallbackUpdate")
    public Cita update(Long id, Cita data, Long medicoId) {
        Cita existing = getById(id);
        if (medicoId != null) {
            existing.setMedico(medicoService.getById(medicoId));
        }
        existing.setEspecialidad(data.getEspecialidad());
        existing.setFechaHora(data.getFechaHora());
        existing.setEstado(data.getEstado());
        return citaRepository.save(existing);
    }

    public Cita fallbackUpdate(Long id, Cita data, Long medicoId, Exception e) {
        throw new RuntimeException("No se puede actualizar la cita. Servicio no disponible temporalmente.");
    }

    @Transactional
    @CircuitBreaker(name = "citasService", fallbackMethod = "fallbackCancelarCita")
    public Cancelacion cancelarCita(Long idCita, String motivo, CanceladoPor por) {
        Cita cita = getById(idCita);

        cita.setEstado(EstadoCita.CANCELADA);
        citaRepository.save(cita);

        Cancelacion cancelacion = Cancelacion.builder()
                .cita(cita)
                .motivo(motivo)
                .canceladoPor(por)
                .fechaCancelacion(LocalDateTime.now())
                .reasignado(false)
                .build();

        cancelacionRepository.save(cancelacion);

        CancelacionEvento evento = new CancelacionEvento();
        evento.setCitaId(cita.getId());
        evento.setPacienteId(cita.getPacienteId());
        evento.setListaEsperaId(cita.getListaEsperaId());
        evento.setMotivo(motivo);

        rabbitTemplate.convertAndSend(RabbitMQConfig.EXCHANGE_NAME, RabbitMQConfig.CANCELACION_ROUTING_KEY, evento);

        return cancelacion;
    }

    public Cancelacion fallbackCancelarCita(Long idCita, String motivo, CanceladoPor por, Exception e) {
        throw new RuntimeException("No se puede cancelar la cita. Servicio no disponible temporalmente.");
    }

    public void delete(Long id) {
        citaRepository.deleteById(id);
    }

    @Data
    public static class CancelacionEvento {
        private Long citaId;
        private Long pacienteId;
        private Long listaEsperaId;
        private String motivo;
    }
}

File: src/services/citaService.js
const { Op } = require('sequelize');
const Cita = require('../models/Cita');
const Cancelacion = require('../models/Cancelacion');
const Medico = require('../models/Medico');
const medicoService = require('./medicoService');
const { publishCancelacion } = require('../config/rabbitmq');

const ESTADOS_VALIDOS = ['PROGRAMADA', 'CANCELADA', 'ATENDIDA', 'REPROGRAMADA'];
const CANCELADO_POR_VALIDO = ['PACIENTE', 'MEDICO', 'ADMINISTRACION'];

const getAll = async () => await Cita.findAll({ include: [{ model: Medico, as: 'medico' }] });

const getById = async (id) => {
  const c = await Cita.findByPk(id, { include: [{ model: Medico, as: 'medico' }] });
  if (!c) { const e = new Error('Cita no encontrada'); e.status = 404; throw e; }
  return c;
};

const getByPacienteId = async (pacienteId) =>
  await Cita.findAll({ where: { pacienteId }, include: [{ model: Medico, as: 'medico' }] });

const getByPacienteIdPaginado = async (pacienteId, page = 0, size = 10) => {
  const offset = page * size;
  const { count, rows } = await Cita.findAndCountAll({
    where: { pacienteId },
    include: [{ model: Medico, as: 'medico' }],
    limit: size,
    offset,
  });
  return {
    content: rows,
    totalElements: count,
    totalPages: Math.ceil(count / size),
    currentPage: page,
  };
};

const getByMedicoId = async (medicoId) =>
  await Cita.findAll({ where: { medicoId }, include: [{ model: Medico, as: 'medico' }] });

const getHorasOcupadas = async (medicoId, fecha) => {
  const inicio = new Date(`${fecha}T00:00:00`);
  const fin = new Date(`${fecha}T23:59:59`);
  const citas = await Cita.findAll({
    where: {
      medicoId,
      fechaHora: { [Op.between]: [inicio, fin] },
      estado: { [Op.ne]: 'CANCELADA' },
    },
  });
  return citas.map(c => {
    const d = new Date(c.fechaHora);
    return `${String(d.getHours()).padStart(2, '0')}:${String(d.getMinutes()).padStart(2, '0')}`;
  });
};

const agendarCita = async (citaData, medicoId) => {
  if (!citaData.pacienteId || !medicoId) {
    const e = new Error('pacienteId y medicoId son obligatorios'); e.status = 400; throw e;
  }
  const medico = await medicoService.getById(medicoId);
  const cita = await Cita.create({
    pacienteId: citaData.pacienteId,
    listaEsperaId: citaData.listaEsperaId,
    medicoId: medico.id,
    especialidad: citaData.especialidad || medico.especialidad,
    fechaHora: citaData.fechaHora,
    estado: 'PROGRAMADA',
  });
  return getById(cita.id);
};

const update = async (id, citaData, medicoId) => {
  const cita = await Cita.findByPk(id);
  if (!cita) { const e = new Error('Cita no encontrada'); e.status = 404; throw e; }
  const updates = {};
  if (medicoId) {
    const medico = await medicoService.getById(medicoId);
    updates.medicoId = medico.id;
  }
  if (citaData.especialidad) updates.especialidad = citaData.especialidad;
  if (citaData.fechaHora) updates.fechaHora = citaData.fechaHora;
  if (citaData.estado && ESTADOS_VALIDOS.includes(citaData.estado)) updates.estado = citaData.estado;
  await cita.update(updates);
  return getById(id);
};

const cancelarCita = async (id, motivo, canceladoPor) => {
  if (!motivo) { const e = new Error('El motivo es obligatorio'); e.status = 400; throw e; }
  if (!CANCELADO_POR_VALIDO.includes(canceladoPor)) {
    const e = new Error(`canceladoPor debe ser: ${CANCELADO_POR_VALIDO.join(', ')}`); e.status = 400; throw e;
  }

  const cita = await Cita.findByPk(id);
  if (!cita) { const e = new Error('Cita no encontrada'); e.status = 404; throw e; }

  await cita.update({ estado: 'CANCELADA' });

  const cancelacion = await Cancelacion.create({
    citaId: cita.id,
    motivo,
    canceladoPor,
    fechaCancelacion: new Date(),
    reasignado: false,
  });

  // Publicar evento en RabbitMQ
  publishCancelacion({
    citaId: cita.id,
    pacienteId: cita.pacienteId,
    listaEsperaId: cita.listaEsperaId,
    motivo,
  });

  return cancelacion;
};

const remove = async (id) => {
  const cita = await Cita.findByPk(id);
  if (!cita) { const e = new Error('Cita no encontrada'); e.status = 404; throw e; }
  await cita.destroy();
};

module.exports = {
  getAll, getById, getByPacienteId, getByPacienteIdPaginado,
  getByMedicoId, getHorasOcupadas, agendarCita, update, cancelarCita, remove,
};
